### PR TITLE
 fix(bazel): Always disable the disk cache 

### PR DESCRIPTION
**Pull request by fviernau**
See individual commits.

### CODE REVIEW

**Comment by fviernau** [comment]
Reading up on [1], it may be better to always disable the disk cache. From the comments I guess that it
speeds up builds only if you make changes to the code / switch branches, which is not the case in ORT analysis. 
Any thoughts @haikoschol @sschuberth ?

[1] https://stackoverflow.com/questions/61282419/bazel-how-outputroot-and-disk-cache-relate-regarding-local-caching

**Comment by nnobelis** [comment]
@fviernau As I mentioned in the issue, you need to specify the parameter (temp disk cache or no disk cache) to EVERY Bazel command run by ORT, even the `--version`.

File: plugins/package-managers/bazel/src/funTest/assets/projects/synthetic/bazel/.bazelrc [hunk 1]
```diff
@@ -0,0 +1 @@
+common --disk_cache=/non-existing-dir
```
Comment (nnobelis) [inline plugins/package-managers/bazel/src/funTest/assets/projects/synthetic/bazel/.bazelrc:1]: Missing line at the end

**Comment by fviernau** [comment]
> @fviernau As I mentioned in the issue, you need to specify the parameter (temp disk cache or no disk cache) to EVERY Bazel command run by ORT, even the `--version`.

Did I miss any bazel command invocations?

File: plugins/package-managers/bazel/src/main/kotlin/Bazel.kt [hunk 2]
```diff
@@ -20,6 +20,7 @@
 package org.ossreviewtoolkit.plugins.packagemanagers.bazel
```
Comment (nnobelis) [inline plugins/package-managers/bazel/src/main/kotlin/Bazel.kt:20]: Commit message: fixes the related issue.

**Comment by nnobelis** [comment]
> > @fviernau As I mentioned in the issue, you need to specify the parameter (temp disk cache or no disk cache) to EVERY Bazel command run by ORT, even the `--version`.
> 
> Did I miss any bazel command invocations?

As I said `--version` here https://github.com/oss-review-toolkit/ort/blob/main/plugins/package-managers/bazel/src/main/kotlin/Bazel.kt#L85

**Comment by haikoschol** [comment]
> Reading up on [1], it may be better to **always** disable the disk cache. From the comments I guess that it speeds up builds only if you make changes to the code / switch branches, which is not the case in ORT analysis. Any thoughts @haikoschol @sschuberth ?

I agree that always disabling the disk cache seems like the best approach.

**Comment by fviernau** [comment]
> As I said `version` here https://github.com/oss-review-toolkit/ort/blob/main/plugins/package-managers/bazel/src/main/kotlin/Bazel.kt#L85

I am able to run `bazel --version` successfully. Have you managed to reproduce the problem with `bazel --version` ?

**Comment by nnobelis** [comment]
This is really strange. `bazel version` works fine in the CLI but it fails when ORT runs it:

```
Exception in thread "main" java.io.IOException: Running 'bazel version' in '/path' failed with exit code 2:
ERROR: /home/dockeruser/.cache/bazel_disk_cache (Permission denied)
ERROR: Error initializing RemoteModule
```

**Comment by nnobelis** [comment]
@fviernau  As a sidenote, please note that there is `bazel version` run by ORT which gives a verbose output and `bazel --version` which just give the version number.

Maybe we can use `--version` and remove `transformBazelVersion` alltogether.

@haikoschol  Would it be ok?

**Comment by sschuberth** [comment]
> `bazel version` works fine in the CLI but it fails when ORT runs it:

Maybe due to different implicit working directories?

**Comment by fviernau** [comment]
> Maybe due to different implicit working directories?

I've run bazel in the directory in which the `.bazelrc` resides in.

File: plugins/package-managers/bazel/src/main/kotlin/Bazel.kt [hunk 3]
```diff
@@ -171,7 +171,10 @@ class Bazel(
         }.getOrNull()
 
     private fun getDependencyGraph(projectDir: File, depDirectives: Map<String, BazelDepDirective>): Set<Scope> {
-        val modGraphProcess = run("mod", "graph", "--output", "json", workingDir = projectDir)
+        val modGraphProcess = run(
+            "mod", "graph", "--output", "json", "--disk_cache", "\"\"", workingDir = projectDir
```
Comment (nnobelis) [inline plugins/package-managers/bazel/src/main/kotlin/Bazel.kt:175]: I did my tests with `run("mod", "graph", "--output", "json", "--disk_cache=", workingDir = projectDir)` and it worked fine.

**Comment by haikoschol** [comment]
> @fviernau As a sidenote, please note that there is `bazel version` run by ORT which gives a verbose output and `bazel --version` which just give the version number.
> 
> Maybe we can use `--version` and remove `transformBazelVersion` alltogether.
> 
> @haikoschol Would it be ok?

I don't see why not. Weird that I didn't discover `--version` myself back then.

BTW, I run Bazel through Bazelisk on my laptop, which seems to intercept the `version` command:
```
$ bazel version
Bazelisk version: 1.20.0
Build label: 7.2.1
Build target: @@//src/main/java/com/google/devtools/build/lib/bazel:BazelServer
Build time: Tue Jun 25 15:53:57 2024 (1719330837)
Build timestamp: 1719330837
Build timestamp as int: 1719330837
```

And `--version` still has some more output than the version number, but again, might be a Bazelisk thing:

```
$ bazel --version
bazel 7.2.1
```

**Comment by fviernau** [comment]
Thanks guys, very helpful! I've added a commit to swich to `bazel --version`.

File: plugins/package-managers/bazel/src/test/kotlin/BazelTest.kt [hunk 4]
```diff
@@ -22,32 +22,13 @@ package org.ossreviewtoolkit.plugins.packagemanagers.bazel
...
 
 import org.ossreviewtoolkit.model.config.AnalyzerConfiguration
 import org.ossreviewtoolkit.model.config.RepositoryConfiguration
 import org.ossreviewtoolkit.utils.test.getAssetFile
 
 class BazelTest : WordSpec({
-    "transformBazelVersion()" should {
-        "remove everything except for the version number" {
-            val bazelVersionOutput = """
-                Bazelisk version: development
```
Comment (haikoschol) [inline plugins/package-managers/bazel/src/test/kotlin/BazelTest.kt:36]: I even put the Bazelisk output in the test, even though I didn't use it in the Docker image. Surprising that this ever worked. 😅

File: plugins/package-managers/bazel/src/main/kotlin/Bazel.kt [hunk 5]
```diff
@@ -82,9 +82,7 @@ class Bazel(
 
     override fun command(workingDir: File?) = "bazel"
 
-    override fun getVersionArguments() = "version"
-
-    override fun transformVersion(output: String) = transformBazelVersion(output)
+    override fun transformVersion(output: String) = output.removePrefix("bazel ")
```
Comment (sschuberth) [inline plugins/package-managers/bazel/src/main/kotlin/Bazel.kt:85]: Nit: Maybe say "Using the *default* `bazel --version`..." to better explain why this work by basically just removing code.

File: plugins/package-managers/bazel/src/main/kotlin/Bazel.kt (hunk 3 above)
Comment (sschuberth) [reply plugins/package-managers/bazel/src/main/kotlin/Bazel.kt:173]: So you mean to remove the "dummy" `"\"\""` argument in favor of adding a `=` after the option name?

Comment (fviernau) [reply plugins/package-managers/bazel/src/main/kotlin/Bazel.kt]: I changed it as @nnobelis has put it.

**Review by sschuberth — APPROVED** [review]

**Pull request merged**